textx/model.py

```diff
@@ -289,6 +301,85 @@ def get_model_from_str(self, model_str, file_name=None, debug=None,
                 pass
             return model
 
+        # Custom attr dunder methods used for user classes during loading
+        def _hasattr(obj, name):
+            try:
+                return name in obj._tx_obj_attrs[id(obj)]
+            except KeyError:
+                if obj._tx_real_hasattr:
+                    return obj._tx_real_hasattr(name)
+                else:
+                    return super(type(obj), obj).__hasattr__(name)
+            except AttributeError:
+                return False
+
+        def _getattr(obj, name):
+            try:
+                return obj._tx_obj_attrs[id(obj)][name]
+            except KeyError:
+                if obj._tx_real_getattr:
+                    return obj._tx_real_getattr(name)
+                else:
+                    return super(type(obj), obj).__getattr__(name)
+
+        def _setattr(obj, name, value):
+            try:
+                obj._tx_obj_attrs[id(obj)][name] = value
+            except KeyError:
+                if obj._tx_real_setattr:
+                    return obj._tx_real_setattr(name, value)
+                else:
+                    return super(type(obj), obj).__setattr__(name, value)
+
+        def _delattr(obj, name):
+            try:
+                obj._tx_obj_attrs[id(obj)].pop(name)
+            except KeyError:
+                if obj._tx_real_delattr:
+                    return obj._tx_real_delattr(name)
+                else:
+                    return super(type(obj), obj).__delattr__(name)
+
+        def _replace_user_attr_methods(self):
+            """
+            Replace has/get/set/del(attr) methods on user classes
+            to support postponing of user obj initialization.
+            """
+            for user_class in self.metamodel.user_classes.values():
+                if not hasattr(user_class, '_tx_instrumented'):
+                    for a_name in ('has', 'get', 'set', 'del'):
+                        real_name = '__{}attr__'.format(a_name)
+                        cached_name = '_tx_real_{}attr'.format(a_name)
+                        setattr(user_class, cached_name,
+                                getattr(user_class, real_name, None))
+                        setattr(user_class, real_name,
+                                getattr(self.__class__,
+                                        '_{}attr'.format(a_name)))
+                    user_class._tx_instrumented = 1
+                else:
+                    user_class._tx_instrumented += 1
```

Comment:
There is a situation where the same user class might  be used in multiple (meta)model (like in test_issue140). Since instrumentation is not idempotent operation we must be sure to apply it only once and count the number of application to revert the operation when the count drops to 0 (i.e. each application has its undo but they are nested). Another approach would be to use a bool flag and instrument only once and do instrumentation removal at the place which comes after all loading is complete (with all imported models) and we are sure that instrumentation is not needed anymore.